src/compiler/crystal/syntax/lexer.cr

```diff
@@ -2524,14 +2525,23 @@ module Crystal
       @token
     end
 
-    def lookahead
-      old_pos = @reader.pos
-      old_line_number, old_column_number = @line_number, @column_number
+    def lookahead(preserve_token_on_fail = false)
+      old_pos, old_line, old_column = current_pos, @line_number, @column_number
+      @temp_token.copy_from(@token) if preserve_token_on_fail
 
       result = yield
       unless result
-        @reader.pos = old_pos
-        @line_number, @column_number = old_line_number, old_column_number
+        self.current_pos, @line_number, @column_number = old_pos, old_line, old_column
+        @token.copy_from(@temp_token) if preserve_token_on_fail
+      end
+      result
+    end
+
+    def peek_ahead
+      result = uninitialized typeof(yield)
+      lookahead(preserve_token_on_fail: true) do
+        result = yield
+        nil
       end
       result
```

Comment:
This can be shortened and simplified:
```suggestion
      lookahead(preserve_token_on_fail: true) do
        yield
      end
```

Reply:
Then if `yield` produces a truthy value, the lexer would not restore to the previous state, but `peek_ahead` is supposed to always do this, so this is incorrect.